src/org/zaproxy/zap/authentication/PostBasedAuthenticationMethodType.java

```diff
@@ -252,6 +260,43 @@ protected HttpMessage prepareRequestMessage(UsernamePasswordAuthenticationCreden
 			return requestMessage;
 		}
 
+		private void replaceAntiCsrfTokenAndCookieValueIfRequired(HttpMessage requestMessage) throws URIException, NullPointerException, HttpMalformedHeaderException {
+			if(extAntiCsrf == null) {
+				extAntiCsrf = Control.getSingleton().getExtensionLoader().getExtension(ExtensionAntiCSRF.class);
+			}
+			List<AntiCsrfToken> antiCsrfTokensOfFirstLoginMsg = null;
+			if(extAntiCsrf != null) {
+				antiCsrfTokensOfFirstLoginMsg = extAntiCsrf.getTokens(requestMessage);
+			}
+			AntiCsrfToken antiCsrfTokenOfFirstLoginMsg = null;
+			if(antiCsrfTokensOfFirstLoginMsg != null && antiCsrfTokensOfFirstLoginMsg.size() > 0) {
+				antiCsrfTokenOfFirstLoginMsg = antiCsrfTokensOfFirstLoginMsg.get(0);
+			}
+			
+			if (antiCsrfTokenOfFirstLoginMsg != null) {
+				this.requireAnitCsrfTokenWhenLogin = true;
+				URI loginURI = new URI(loginRequestURL, true, "UTF-8");
+				HttpMessage loginMsgToRegenerateAntiCsrfToken = new HttpMessage(loginURI);
+				loginMsgToRegenerateAntiCsrfToken.setRequestingUser(null);
+				try {
+					getHttpSender().sendAndReceive(loginMsgToRegenerateAntiCsrfToken);
+				} catch (IOException e) {
```

Comment:
Should not catch it? If the request requires a fresh anti-csrf token there's no point trying to authenticate.

Reply:
Actually, I logged out to perform testing. When logging out the application redirects to the login page with a new token.
With the `loginSiteNode` I could not get the new token. It always gives the very first token sent with the post request to login site(when authenticating using browser).
That's why I make a new request and get the token from that.
BTW, I do not understand why there's no point in trying to authenticate.

Reply:
Yes, the `loginSiteNode` does not need to be used for the regeneration of the token, we should use the message from the `AntiCsrfToken` (which has the correct URL).

Because the authentication will fail (why try it anyway?).